File: jobs/2026-05-27__14-51-17/auto_generate_extract_schema_py__dtEWFz4/artifacts/user/extract_task/extract.py
#!/usr/bin/env python3
"""
Auto-generate a LlamaExtract Schema and Run Extraction

This script:
1. Uploads an invoice PDF to LlamaCloud with a unique external_file_id
2. Auto-generates a JSON Schema for invoice data
3. Uses the generated schema to run structured extraction
4. Saves the schema, extraction result, and log files
"""

import os
import json
import time
from llama_cloud import LlamaCloud

def main():
    # Get environment variables
    api_key = os.environ.get("LLAMA_CLOUD_API_KEY")
    if not api_key:
        raise ValueError("LLAMA_CLOUD_API_KEY environment variable is not set")
    
    run_id = os.environ.get("ZEALT_RUN_ID")
    if not run_id:
        raise ValueError("ZEALT_RUN_ID environment variable is not set")
    
    # Initialize the LlamaCloud client
    client = LlamaCloud(api_key=api_key)
    
    # Define file paths
    pdf_path = "/home/user/extract_task/data/invoice.pdf"
    schema_path = "/home/user/extract_task/schema.json"
    result_path = "/home/user/extract_task/result.json"
    log_path = "/home/user/extract_task/output.log"
    
    # Upload the PDF file with unique external_file_id
    external_file_id = f"invoice-{run_id}.pdf"
    print(f"Uploading file with external_file_id: {external_file_id}")
    
    try:
        with open(pdf_path, "rb") as f:
            uploaded_file = client.files.create(
                file=f,
                purpose="extract",
                external_file_id=external_file_id
            )
        file_id = uploaded_file.id
        print(f"File uploaded successfully. File ID: {file_id}")
    except Exception as e:
        # File might already exist, try to find it
        print(f"Upload failed (file may already exist): {e}")
        print("Trying to find existing file...")
        
        # List files to find the one with matching external_file_id
        files = client.files.list()
        for file in files:
            if hasattr(file, 'external_file_id') and file.external_file_id == external_file_id:
                file_id = file.id
                print(f"Found existing file with ID: {file_id}")
                break
        else:
            # If not found, try with a different external_file_id
            timestamp = int(time.time())
            external_file_id = f"invoice-{run_id}-{timestamp}.pdf"
            print(f"Trying with new external_file_id: {external_file_id}")
            with open(pdf_path, "rb") as f:
                uploaded_file = client.files.create(
                    file=f,
                    purpose="extract",
                    external_file_id=external_file_id
                )
            file_id = uploaded_file.id
            print(f"File uploaded successfully. File ID: {file_id}")
    
    # Generate schema for invoice data
    print("Generating schema for invoice data...")
    generated_schema = client.extract.generate_schema(
        prompt="Extract structured invoice data including invoice number, vendor/supplier information, line items, dates, and total amounts",
        file_id=file_id
    )
    
    # Extract the schema from the response
    schema = generated_schema.parameters.data_schema
    
    # Save the generated schema to disk
    print(f"Saving generated schema to {schema_path}")
    with open(schema_path, "w") as f:
        json.dump(schema, f, indent=2)
    
    # Extract top-level property names from the schema
    if "properties" in schema:
        property_names = list(schema["properties"].keys())
    else:
        property_names = []
    
    print(f"Schema generated with {len(property_names)} properties")
    
    # Create extraction configuration using the generated schema
    print("Creating extraction job...")
    extraction_config = {
        "extraction_target": "per_doc",
        "tier": "agentic",
        "data_schema": schema
    }
    
    extraction_job = client.extract.create(
        configuration=extraction_config,
        file_input=file_id
    )
    
    job_id = extraction_job.id
    print(f"Extraction job created. Job ID: {job_id}")
    
    # Poll until the job reaches a terminal state
    print("Polling for job completion...")
    terminal_states = ["COMPLETED", "FAILED", "CANCELLED"]
    
    while True:
        job_status = client.extract.get(job_id)
        status = job_status.status
        
        print(f"Current status: {status}")
        
        if status in terminal_states:
            break
        
        time.sleep(2)  # Wait 2 seconds before polling again
    
    final_status = job_status.status
    print(f"Job finished with status: {final_status}")
    
    # Get the extraction result if job completed successfully
    if final_status == "COMPLETED":
        if hasattr(job_status, 'extract_result') and job_status.extract_result:
            extraction_result = job_status.extract_result
        else:
            # Try to get the result from the job's result field
            extraction_result = job_status.result
    else:
        extraction_result = {"error": f"Job failed with status: {final_status}"}
    
    # Save the extraction result to disk
    print(f"Saving extraction result to {result_path}")
    with open(result_path, "w") as f:
        json.dump(extraction_result, f, indent=2)
    
    # Write log file with required information
    print(f"Writing log file to {log_path}")
    with open(log_path, "w") as f:
        # Schema fields line
        f.write(f"Schema fields: {', '.join(property_names)}\n")
        # Job ID line
        f.write(f"Job ID: {job_id}\n")
        # Status line
        f.write(f"Status: {final_status}\n")
    
    print("Extraction process completed successfully!")
    print(f"Schema saved to: {schema_path}")
    print(f"Result saved to: {result_path}")
    print(f"Log saved to: {log_path}")
    
    # Verify the log file contents
    print("\nVerifying log file contents:")
    with open(log_path, "r") as f:
        for line in f:
            print(f"  {line.rstrip()}")

if __name__ == "__main__":
    main()

File: jobs/2026-07-02__07-42-24/llamacloud_per_table_row_extract__SzV6fLs/artifacts/user/myproject/extract.ts
import fs from 'fs';
import z from 'zod/v4';
import { LlamaCloud } from '@llamaindex/llama-cloud';

async function main() {
  // 1. Read run ID
  const runIdPath = '/logs/artifacts/run-id';
  if (!fs.existsSync(runIdPath)) {
    throw new Error(`Run ID file not found at ${runIdPath}`);
  }
  const runId = fs.readFileSync(runIdPath, 'utf8').trim();
  console.log(`Using run ID: ${runId}`);

  // 2. Verify API key
  const apiKey = process.env.LLAMA_CLOUD_API_KEY;
  if (!apiKey) {
    throw new Error('LLAMA_CLOUD_API_KEY environment variable is not set');
  }

  // 3. Initialize LlamaCloud client
  const client = new LlamaCloud({ apiKey });

  // 4. Define Zod schema and convert to JSON Schema
  const schema = z.object({
    product_code: z.string(),
    product_name: z.string(),
    category: z.string(),
    price_usd: z.number(),
    stock: z.number().int(),
  });

  const data_schema = z.toJSONSchema(schema);
  console.log('JSON Schema generated:', JSON.stringify(data_schema, null, 2));

  // 5. Upload source PDF
  const pdfPath = '/home/user/myproject/data/products.pdf';
  if (!fs.existsSync(pdfPath)) {
    throw new Error(`PDF file not found at ${pdfPath}`);
  }

  console.log('Uploading PDF...');
  const fileStream = fs.createReadStream(pdfPath);
  const fileUpload = await client.files.create({
    file: fileStream,
    purpose: 'extract',
    external_file_id: `products-${runId}.pdf`,
  });

  const fileId = fileUpload.id;
  console.log(`Uploaded file ID: ${fileId}`);

  // 6. Submit extraction job
  console.log('Submitting extraction job...');
  let job = await client.extract.create({
    file_input: fileId,
    configuration: {
      data_schema: data_schema as any,
      extraction_target: 'per_table_row',
      tier: 'agentic',
    },
  });

  const jobId = job.id;
  console.log(`Job ID: ${jobId}, Initial status: ${job.status}`);

  // 7. Poll until terminal status
  const terminalStatuses = ['COMPLETED', 'FAILED', 'CANCELLED'];
  while (!terminalStatuses.includes(job.status)) {
    console.log(`Polling... current status: ${job.status}`);
    await new Promise((resolve) => setTimeout(resolve, 2000));
    job = await client.extract.get(jobId);
  }

  console.log(`Job finished with status: ${job.status}`);

  if (job.status !== 'COMPLETED') {
    throw new Error(`Job failed or cancelled: ${job.error_message || 'No error message provided'}`);
  }

  // 8. Write extraction result to output.json
  const result = job.extract_result;
  if (!result) {
    throw new Error('Completed job has no extract_result');
  }

  const outputPath = '/home/user/myproject/output.json';
  fs.writeFileSync(outputPath, JSON.stringify(result, null, 2), 'utf8');
  console.log(`Wrote output to ${outputPath}`);

  // 9. Append a single log line to output.log
  const logPath = '/home/user/myproject/output.log';
  const numRows = Array.isArray(result) ? result.length : 0;
  fs.appendFileSync(logPath, `Extracted rows: ${numRows}\n`, 'utf8');
  console.log(`Appended log to ${logPath}`);
}

main().catch((err) => {
  console.error('Error:', err);
  process.exit(1);
});
